### PR TITLE
[BUGFIX] Auto calculation of listing heights + dropup menus

### PR DESCRIPTION
**Pull request by johncowen**
Addresses #4125

All table listings now auto calculate their own height depending on the space left in the page.

Additionally, the dropdown menus now detect whether they will be cut off by the 'masked container' and will now 'drop up' if so.

### CODE REVIEW

File: ui-v2/app/components/tabular-collection.js [hunk 1]
```diff
@@ -37,6 +56,21 @@ const change = function(e) {
     const value = e.currentTarget.value;
     if (value != get(this, 'checked')) {
       set(this, 'checked', value);
+      if (e.currentTarget.getAttribute('id') !== 'actions_close') {
```
Comment (meirish) [inline ui-v2/app/components/tabular-collection.js]: So this keeps component state on the DOM via an id of `actions_close`?

File: ui-v2/app/components/tabular-collection.js [hunk 2]
```diff
@@ -37,6 +56,21 @@ const change = function(e) {
...
+        const footerRect = $footer.getBoundingClientRect();
+        const groupBottom = groupRect.top + $group.clientHeight;
+        const footerTop = footerRect.top;
+        if (groupBottom > footerTop) {
+          $group.classList.add('above');
+        } else {
+          $group.classList.remove('above');
+        }
+        $tr.style.zIndex = maxZIndex + 1;
+      }
```
Comment (meirish) [inline ui-v2/app/components/tabular-collection.js]: When does this `change` fn fire?

File: ui-v2/app/components/tabular-collection.js [hunk 3]
```diff
@@ -86,12 +120,15 @@ export default Component.extend(SlotsMixin, {
     window.removeEventListener('resize', this.handler);
   },
   resize: function(e) {
-    const $footer = [...$$('#wrapper > footer')][0];
-    const $thead = [...$$('main > div')][0];
-    if ($thead) {
-      // TODO: This should auto calculate properly from the CSS
-      this.set('height', Math.max(0, new Number(e.detail.height - ($footer.clientHeight + 218))));
-      this['cell-layout'] = new ZIndexedGrid($thead.clientWidth, 50);
+    const $tbody = this.$('tbody').get(0);
```
Comment (meirish) [inline ui-v2/app/components/tabular-collection.js]: I _think_ you could do `$$('tbody', this.element)[0]` here if you didn't want to mix 'n match jQuery & DOM stuff.

File: ui-v2/app/components/tabular-collection.js [hunk 4]
```diff
@@ -106,6 +143,108 @@ export default Component.extend(SlotsMixin, {
...
+      if (priorMap) {
+        cell = priorMap[itemKey];
+      }
+      if (cell) {
+        // additional `checked` argument
+        style = this._cellLayout.formatItemStyle(
+          itemIndex,
+          this._clientWidth,
+          this._clientHeight,
+          this.checked
```
Comment (meirish) [inline ui-v2/app/components/tabular-collection.js]: is `checked` the "You've clicked the menu on this item" ? state?

File: ui-v2/app/components/tabular-collection.js (hunk 1 above)
Comment (johncowen) [reply ui-v2/app/components/tabular-collection.js]: You probably need a little more context to follow this, feel free to ping me and I can walk you through it.

TLDR; It doesn't run the code/fork if the selected radio button on change is the 'reset radiogroup' (`id="actions_close"`) button.

Basically there is an `<input type="radio" id="actions_close" />` that is hidden in the `<th>` of the 'actions column' in the table. All of the other dropdowns use the same radiobutton `name=""`/radiogroup. This ensures you can only ever have one dropdown visible at once, and I have a way to 'unset' the group (the 'actions_close' radio essentially 'resets' the radiogroup). 

The above code needs to run whenever a dropdown menu is opened, in other words, whenever the radiogroup selection changes. The thing is, the 'action_close' radio is in the group also, so if the action_close button was selected/clicked (in other words I didn't open a dropdown, but reset the group/ensured they are all closed), don't run the code.

Sorry probably not the best explanation, ping me if you want more info.

File: ui-v2/app/components/tabular-collection.js (hunk 2 above)
Comment (johncowen) [reply ui-v2/app/components/tabular-collection.js]: When a dropdown is opened (under the hood, this is when a radiobutton is selected).

File: ui-v2/app/components/tabular-collection.js (hunk 3 above)
Comment (johncowen) [reply ui-v2/app/components/tabular-collection.js]: 👍 Ah nice yeah, will try that

File: ui-v2/app/components/tabular-collection.js (hunk 4 above)
Comment (johncowen) [reply ui-v2/app/components/tabular-collection.js]: Ah no, I see what it looks like :) but it's not a boolean, it's the index of the checked 'cell'. I think this was the 'best' way I could think of to fiddle with as little of `ember-collection` as possible.

I already have a custom `ZIndexedGrid` up above, so I'm passing the 'checked index' through here in 3 places, and I've done the logic up in in one place in my already custom `ZIndexedGrid`. I commented up the 3 places where I had to change the `ember-collection` code by passing through an extra arg also.

File: ui-v2/app/components/tabular-collection.js [hunk 5]
```diff
@@ -106,6 +143,108 @@ export default Component.extend(SlotsMixin, {
...
+        this._clientWidth,
+        this._clientHeight,
+        this.checked
+      );
+      cell = new Cell(itemKey, item, itemIndex, style);
+      cellMap[itemKey] = cell;
+      this._cells.pushObject(cell);
+    }
+    this._cellMap = cellMap;
+  },
```
[the remark this answers is not in the document]
Comment (johncowen) [reply ui-v2/app/components/tabular-collection.js]: Yeah for sure, very good point. It's not massively clear whats going on here, will try and add a bit more context in the comments 👍

[the remark this answers is not in the document]
Comment (meirish) [reply ui-v2/app/components/tabular-collection.js]: Yep, the added comments help a lot - thanks!

File: ui-v2/app/components/tabular-collection.js (hunk 4 above)
Comment (meirish) [reply ui-v2/app/components/tabular-collection.js]: Ahh I thought `checked` like the form element state, not an index.
